device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -267,64 +291,91 @@ public void onConnectionEstablished(String connectionId)
             this.updateStatus(IotHubConnectionStatus.CONNECTED, IotHubConnectionStatusChangeReason.CONNECTION_OK, null);
         }
     }
 
+    @Override
+    public void onMultiplexedDeviceSessionEstablished(String connectionId, String deviceId)
+    {
+        if (connectionId.equals(this.iotHubTransportConnection.getConnectionId()))
+        {
+            log.debug("The device session in the multiplexed connection to the IoT Hub has been established for device {}", deviceId);
+            this.updateStatus(IotHubConnectionStatus.CONNECTED, IotHubConnectionStatusChangeReason.CONNECTION_OK, null, deviceId);
+        }
+    }
+
+    @Override
+    public void onMultiplexedDeviceSessionLost(Throwable e, String connectionId, String deviceId)
+    {
+        if (connectionId.equals(this.iotHubTransportConnection.getConnectionId()))
+        {
+            log.debug("The device session in the multiplexed connection to the IoT Hub has been lost for device {}", deviceId);
+            if (e == null)
+            {
+                this.updateStatus(IotHubConnectionStatus.DISCONNECTED, IotHubConnectionStatusChangeReason.CLIENT_CLOSE, null, deviceId);
+            }
+            else
+            {
+                this.updateStatus(IotHubConnectionStatus.DISCONNECTED_RETRYING, exceptionToStatusChangeReason(e), e, deviceId);
+
+                if (e instanceof TransportException)
+                {
+                    this.reconnectDeviceSession((TransportException) e, deviceId);
+                }
+                else
+                {
+                    this.reconnectDeviceSession(new TransportException(e), deviceId);
+                }
+            }
+        }
+    }
+
+    public void setMultiplexingRetryPolicy(RetryPolicy retryPolicy)
+    {
+        this.multiplexingRetryPolicy = retryPolicy;
+    }
+
     /**
      * Establishes a communication channel with an IoT Hub. If a channel is
      * already open, the function shall do nothing.
      *
      * If reconnection is occurring when this is called, this function shall block and wait for the reconnection
      * to finish before trying to open the connection
      *
-     * @param deviceClientConfigs the configs for the devices to open
-     *
      * @throws DeviceClientException if a communication channel cannot be
      * established.
      */
-    public void open(Collection<DeviceClientConfig> deviceClientConfigs) throws DeviceClientException
+    public void open() throws DeviceClientException
     {
-        if ((deviceClientConfigs == null) || deviceClientConfigs.isEmpty())
-        {
-            //Codes_SRS_IOTHUBTRANSPORT_34_015: [If the provided list of configs is null or empty, this function shall
-            // throw an IllegalArgumentException.]
-            throw new IllegalArgumentException("deviceClientConfigs cannot be null or empty");
-        }
-
         if (this.connectionStatus == IotHubConnectionStatus.CONNECTED)
```

Comment:
Is the call to Open() thread safe?

Reply:
Not at this level, but it is at the MultiplexingClient level